Fix part of #10474: Cover InteractionSpecsService with strict checks

17sushmita opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #10474 .
2. This PR does the following: Covers interaction-specs service with strict typescript checks

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

FareesHussain committed:
Fix #9427: Remove email access in admin-page (#11869)

Commit:
fix frontend test failing flakily due to coverage (#11907)

savithakj committed:
Fix part of #8472: Migrate Collection Update Service (#11663)

* migrate collection-update

* updated based on linter checks

* updated based on linter checks

* updated based on linter checks

* changed based on feedback

Nik-09 committed:
Fix part of #9749: Migrate admin-prod-mode-activities-tab directive to angular component (#11787)

Nik-09 committed:
Fix Part of #4057: Adds test for some methods (#11863)

Nik-09 committed:
Fix Part of #10700: Convert Object factory to Model Class (#11902)

Mohitbalwani26 committed:
Fix part of #10306: Added async keyword to asynchronous functions that return a promise (#11915)

* added async keyword in exploration-permissions-backend-api.service

* added async keyword in exploration-permissions-backend-api.service

gp201 committed:
Fix #11885: added media query to change to mobile view (#11894)

* Wipeout 8.16: Save the username of deleted user so it cannot be reused (#11168)

* Fix CommunityContributionStatsModel audit job issues (#11171)

* Fix part of #10474: Add strict checks to files (#11028)

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Fix #11032: Adds customization args content-id for state color status for voiceover (#11094)

* Add one-off job for question suggestion migration (#11169)

* Wipeout 9.1: Add tests for wipeout and modify the release for wipeout (#10830)

* Revert " Remove @kevintab95 from CODEOWNERS temporarily (#11174)" (#11189)

This reverts commit cdc45e714db3db34ae43d0b7671527f400b41628.

* Fix #3341: Added confirmation modal in the customization args editing. (#11164)

* Added confirmation modal for customization args.

* Added tests.

* Fixed linting errors.

* Fix part of #10415: Extract config validators from prod_validator (#11148)

* Increase taskqueue http request timeout (#11195)

* Increase taskqueue timeout to 30s

* add import statement

* add a comment

* fix typo

* fix backend test

* Fixes part of #10415: Topic validator refractor  (#10967)

* validators: Refactor topic and subtopic validators into individual validators

* validators: Refactor topic and resolve backend test coverage

* validators: Remove recommendations validator from topic validators file

* validators: Remove unused variable

* validators: Fix docstring naming and update copyright date

* validators: Define model vars in alphabetical order

* validators: Include topic test changes from oppia/oppia#10981

* Fix topic validators

* Fix backend tests

* Fix tests

Co-authored-by: Matt Higgins <matt@matthiggins.dev>
Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Fix part of #8472: Migrate translation language service to Angular (#11178)

* migrated the service

* minor changes

* minor changes

* removed from the upgraded service

* newline change

* Fix Part of #8472: Migrate Topic Rights Backend API Service (#11177)

* Migrate Topic Rights Backend API to angular

* Remove unecessary comments

* Fix minor issues

* Add rootScope

* Add .()

* Issue fixed

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>
Co-authored-by: Sandeep Dubey <dubeysandeep.in@gmail.com>
Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>
Co-authored-by: Kevin Thomas <kevintab@tutanota.com>
Co-authored-by: Prayush Dawda <35144226+iamprayush@users.noreply.github.com>
Co-authored-by: Ramin Izadpanah <ramizd@gmail.com>
Co-authored-by: Matt Higgins <mrhiggi@g.clemson.edu>
Co-authored-by: Matt Higgins <matt@matthiggins.dev>
Co-authored-by: arpit1912 <46648172+arpit1912@users.noreply.github.com>
Co-authored-by: Siddharth Batra <sbsiddharth@gmail.com>

tushar2407 committed:
files added in tsconfig-strict.json with bug fixes in spec.ts (#11878)

lkbhitesh07 committed:
Fix part of #11431: Replace run_in_transaction call with decorator (#11778)

seanlip committed:
Fix part of #11433: Add new TranslatableHtml and TranslatableUnicode objects. (#11937)

ARKEOLOGIST committed:
Fixes #11919 : Refactor CSS styles to prevent errors from looking clickable (#11940)

* Fixed topic editor warnings appearing clickable

* Fixed linting issue

* More linting fixes

krishnarao22 committed:
Adding waitFor for nextCardButton (#11872)

seanlip committed:
Log process info before terminating or killing. (#11938)

Commit:
Add hi-en to SUPPORTED_CONTENT_LANGUAGES (#11917)

Radesh-kumar committed:
Fix part of #8472: Migrate story-update.service (#11889)

* initial

* updated

* updated

* update

* updated

* updated

krishita30j committed:
Fixes issue #11898: The social media icons are missaligned and different sizes (#11913)

* Fixed social media icons

* Modified file

gp201 committed:
Fix #11918: Added text to that conveys the interaction used (#11941)

* message added

* lint fixed

* lint fixes

* lint fixes

* fixed issues

* alt fix

* lint fix

* fixed issues

* used ng-if

17sushmita committed:
Fix part of #10474: Cover InteractionSpecsService with strict checks

U8NWXD committed:
Fix #9456 (macOS SSL Errors): Use Certify for CA Roots (#11107)

* Add the certifi package as a dependency

Certifi provides Mozilla's root certificate store so that we don't have
to rely on the root certificates on a developer's machine or in their
Python installation, which on macOS are often problematic.

* Use certifi certificates in python_utils

When accessing URLs with python_utils `url_open` or `url_retrieve`, use
the root certificates from certifi.

* Add certifi to app_dev.yaml

* Install certify as prerequisite for all scripts

* Fix installation and usage of certifi

* Lint fixes

* Fix backend test

* Stop installing certifi in third_party

* Move common certifi and ssl imports to top of python_utils.py

* Fix backend test

krishita30j committed:
Switching order of panes in the history view (#11854)

krishnarao22 committed:
Uploading Protractor Screenshots as Artifacts for Github Actions suites (#11953)

17sushmita commented:
@vojtechjelinek PTAL

vojtechjelinek commented on core/templates/domain/exploration/InteractionObjectFactory.ts:
```diff
@@ -78,7 +79,7 @@ export interface InteractionBackendDict {
   'customization_args': InteractionCustomizationArgsBackendDict;
   'hints': readonly HintBackendDict[];
-  'id': string;
+  'id': keyof typeof InteractionSpecsConstants.INTERACTION_SPECS;
```
Should we add the `InteractionId` above and then use it everywhere?
```typescript
export type InteractionId = (
  keyof typeof InteractionSpecsConstants.INTERACTION_SPECS)
```

17sushmita replied:
yes, that would be better. I've added the changes

vojtechjelinek commented on core/templates/pages/exploration-player-page/services/content-translation-manager.service.spec.ts:
```diff
@@ -322,6 +325,12 @@ describe('Content translation manager service', () => {
       writtenTranslations.addWrittenTranslation(
         'ca_1', 'fr', 'unicode', 'fr 2');
 
+      // This throws "Type '"DummyInteraction"' is not assignable to type
+      // '"ImageClickInput" | "NumberWithUnits" | "NumericExpressionInput" |
+      // "NumericInput" | "DragAndDropSortInput" | "ItemSelectionInput" |
+      // "Continue" | "GraphInput" | "EndExploration" | ... 11 more ... |
+      // "RatioExpressionInput"'.""
```
```suggestion
      // This throws "Type '"DummyInteraction"' is not assignable to type
      // '"ImageClickInput" | "NumberWithUnits" | "NumericExpressionInput" |
      // "NumericInput" | "DragAndDropSortInput" | "ItemSelectionInput" |
      // "Continue" | "GraphInput" | "EndExploration" | ... 11 more ... |
      // "RatioExpressionInput"'." We need to suppress this error because
      // we need to change the value of 'interaction.id' for testing
      // purposes.
```

vojtechjelinek commented on core/templates/services/interaction-specs.service.spec.ts:
```diff
@@ -29,6 +29,12 @@ describe('InteractionSpecsService', () => {
 
   describe('checking whether an interaction can be trained with ML', () => {
     it('should throw an error when interaction does not exist.', () => {
+      // This throws "Argument of type '"Fake"' is not assignable to parameter
+      // of type '"ImageClickInput" | "NumberWithUnits" |
+      // "NumericExpressionInput" | "NumericInput" | "DragAndDropSortInput" |
+      // "ItemSelectionInput" | "Continue" | "GraphInput" | "EndExploration" |
+      // ... 11 more ... | "RatioExpressionInput"'."
```
```suggestion
      // This throws "Argument of type '"Fake"' is not assignable to parameter
      // of type '"ImageClickInput" | "NumberWithUnits" |
      // "NumericExpressionInput" | "NumericInput" | "DragAndDropSortInput" |
      // "ItemSelectionInput" | "Continue" | "GraphInput" | "EndExploration" |
      // ... 11 more ... | "RatioExpressionInput"'." We need to suppress this 
      // error because we need to change the value of 'interaction.id' for 
      // testing purposes.
```

vojtechjelinek reviewed (changes requested):
Thanks! Looks pretty good, I have left a few comments.

vojtechjelinek committed:
Upgrade libraries and enable TS checks for Google Analytics (#11936)

17sushmita committed:
Added type definition for InteractionID

17sushmita committed:
Update error description in content-translation-manager.service.spec.ts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

17sushmita committed:
Update error description in interaction-specs.service.spec.ts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

17sushmita commented:
@vojtechjelinek PTAL

17sushmita committed:
Resolved lint check errors

U8NWXD committed:
Fix Part of #11496: Lint Check for E2E Action Functions (#11734)

* Add lint check for e2e action scripts

* Add exclude list for end-to-end-action-checks lint test

* Fix end-to-end-action-checks to exclude by base filename

* Shorten action checks rule name

* Fix e2e-action lint failure

* Limit e2e-action checks to e2e files

* Move e2e-action file include/exclude lists to .eslintrc

* Add files to e2e-action exclude list

* Remove Lint Check for action.clear

Rich text editors have a .clear() function, so we cannot use a lint
check to find cases where we want to use action.clear instead.

* Remove checks for clear from e2e-action.spec.js

* Use function name as messageId in e2e-action lint check

* Remove mistakenly committed tmp.py file

* Remove unnecessary newline from .eslintrc

* Simplify e2e-action lint logic

* Replace elem.sendKeys with action.select2

* Remove click before select2

EricZLou committed:
Fix image preload bug (#11964)

vojtechjelinek reviewed (approved):
LGTM! Thanks a lot.

OBITORASU committed:
Fix part of #11431: Replace run_in_transaction call with decorator (#11821)

* Add run_in_transaction decorator

* Add run_in_transaction decorator

* define non-transactional versions of put_multi and delete_multi

* fix backend error for delete_multi

* Replace put_multi parameter from Models to entities

* Improve transactional function docstrings

* improve documentation of put_multi_transactional

* Fix docstring for put_multi_transactional

vojtechjelinek committed:
Fix signup process for user marked as deleted (#11946)

Hudda committed:
Move _check_extra_js_files to the general linter (#11864)

* Move extra js file check

* Fix check logic

* Addressed review comments

srijanreddy98 committed:
Remove Mariana from codeowners file  (#11968)

srijanreddy98 committed:
Fix part of #9749: Migrate all the continue interaction related directives to angular components (#11935)

* Migrate Conitnue interaction

* Add codeowners

* Rename files in check frontend coverage

* Fix e2e?

* Add Continue to component list

* fix count of files

* Update extensions/extensions.module.ts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* e2e tests

* Add todo

* Review

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

DubeySandeep commented on core/templates/pages/exploration-player-page/services/content-translation-manager.service.spec.ts:
```diff
@@ -322,6 +324,14 @@ describe('Content translation manager service', () => {
+      // we need to change the value of 'interaction.id' for testing
+      // purposes.
+      // @ts-expect-error
       interaction.id = 'DummyInteraction';
```
is there any issue with having a correct interactionId here?

17sushmita replied:
yes, as I defined the type definition for interactionID, it can no longer be assigned to a string type that is not in the key of InteractionSpecsConstants.INTERACTION_SPECS. So, this statement will throw an error if we try to assign 'DummyInteraction'.
And this is a unit test for custom Interaction_specs cases, so we cannot add correct interactionID. And I tried to remove this test as we don't need it now, but this reduced the coverage and I was not able to push the changes as the error - `content-translation-manager.service.ts seems to be not completely tested. Make sure it's fully covered.` showed up.

DubeySandeep replied:
>  this is a unit test for custom Interaction_specs cases, so we cannot add correct interactionID. 

I'm not sure why we can't add a correct interactionID here? Can you explain this a bit more? :)

17sushmita replied:
As the test is intended to test the custom interaction, so, I think we don't need this test now but when I removed it, the frontend test coverage was reduced. So I didn't remove the test and suppress this error. I may have misinterpreted the test. Can you explain a bit about this test?

17sushmita replied:
After giving a deep look at this test, I am concluding that the test is intended to test customized args on the interaction_specs. So the correct interaction id can be assigned with customized args it to. Please confirm if I conclude it right?

DubeySandeep replied:
Thanks for looking into it @17sushmita. 

>Please confirm if I conclude it right?

I also think that's correct, I would suggest adding a correct interactionId!

17sushmita replied:
Okay, thanks for the guidance @DubeySandeep . I'm making a commit with all required changes.

DubeySandeep commented on extensions/interactions/rule-input-defs.ts:
```diff
@@ -265,3 +266,6 @@ export type InteractionRuleInputs = (
   SetInputRuleInputs |
   TextInputRuleInputs);
+
+export type InteractionId = (
```
any specific reason for defining it inside rule-input-defs? (We have rule-input-defs, answer-input-defs, should we add it inside a new file interaction-defs?[not sure!])

17sushmita replied:
I was unsure where to define this, and the rule-input-defs was defining the rules for keys of InteractionSpecsConstants.INTERACTION_SPECS, so I defined it along with them. Should I make a new file?

iamprayush replied:
Yes, I also think it makes sense to move it to a separate file.

17sushmita replied:
@iamprayush Can you guide me where should I place this file?

iamprayush replied:
You can create a new file in this directory itself, just like `rule-input-defs.ts`, `answer-defs.ts`, and `customization-args-defs.ts`.

17sushmita replied:
Okay, I'm adding the commit with required changes.

Rijuta-s committed:
Fix Classroom Page : Fixes Classroom page navbar in mobile view. (#11873)

* adding topic name

* changing story viewer navbar according to mocks

* fixing minor errors

* fixing lint issues

* fix linting issues

* fixing lint issues

* minor lint issue

* minor lint fix

* minor fix

* minor issue

* fixing lint

* minor lint -issue

* minor fix

DubeySandeep commented on extensions/interactions/interaction-attributes-extractor.service.ts:
```diff
@@ -25,6 +25,7 @@ import { InteractionCustomizationArgs } from
   'extensions/interactions/customization-args-defs';
 import { InteractionObjectFactory } from
   'domain/exploration/InteractionObjectFactory';
+import { InteractionId } from './rule-input-defs';
```
Can we have the absolute import here? (If not, then add a new line above to separate relative imports from others.)

17sushmita replied:
Yes, we can have it. Thanks for pointing it. I'm adding a commit for this.

DubeySandeep reviewed (changes requested):
@17sushmita Thanks for the PR, the changes look great! :) I've left a few minor comments/questions PTAL!

Rijuta-s committed:
Fix #11887: fixes audio player position in tablet,mobile view (#11948)

* fixing audio bar position in mobile view

* fixing audio header position

* fixing audioplayer

* fixing lint issues

* fixing absolute positioning of audio-bar in preview player

* final changes

* fixing lint issues

17sushmita committed:
Add absolute import for interactionID in interaction-attributes-extractor service

17sushmita commented:
@DubeySandeep PTAL

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks!

sagangwee committed:
Fix part of #11900: Disable opportunities list next button if there are no more results to display (#11982)

* Qualify more check

* Fix linting

* Add TODO

seanlip committed:
Fix part of #11950: Add jobs for migrating all snapshots of explorations and questions to the latest schema version. (#11952)

* Initial audit job.

* Fix schema v33 issue.

* Add the actual migration job.

* Handle the cases where there is no state schema version.

* Fix typo.

* Add one-off jobs for questions too.

* Fix some errors.

* Add missing return statements.

* Fix more issues.

* Fix more issues.

* Add tests.

* Achieve full test coverage.

* Fix a test.

* Fix one more issue.

* Add explanatory comments.

* Remove unneeded asserts.

gp201 committed:
Fix #11912: Implements delay before cards can be rearranged. (#11931)

* Wipeout 8.16: Save the username of deleted user so it cannot be reused (#11168)

* Fix CommunityContributionStatsModel audit job issues (#11171)

* Fix part of #10474: Add strict checks to files (#11028)

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Fix #11032: Adds customization args content-id for state color status for voiceover (#11094)

* Add one-off job for question suggestion migration (#11169)

* Wipeout 9.1: Add tests for wipeout and modify the release for wipeout (#10830)

* Revert " Remove @kevintab95 from CODEOWNERS temporarily (#11174)" (#11189)

This reverts commit cdc45e714db3db34ae43d0b7671527f400b41628.

* Fix #3341: Added confirmation modal in the customization args editing. (#11164)

* Added confirmation modal for customization args.

* Added tests.

* Fixed linting errors.

* Fix part of #10415: Extract config validators from prod_validator (#11148)

* Increase taskqueue http request timeout (#11195)

* Increase taskqueue timeout to 30s

* add import statement

* add a comment

* fix typo

* fix backend test

* Fixes part of #10415: Topic validator refractor  (#10967)

* validators: Refactor topic and subtopic validators into individual validators

* validators: Refactor topic and resolve backend test coverage

* validators: Remove recommendations validator from topic validators file

* validators: Remove unused variable

* validators: Fix docstring naming and update copyright date

* validators: Define model vars in alphabetical order

* validators: Include topic test changes from oppia/oppia#10981

* Fix topic validators

* Fix backend tests

* Fix tests

Co-authored-by: Matt Higgins <matt@matthiggins.dev>
Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Fix part of #8472: Migrate translation language service to Angular (#11178)

* migrated the service

* minor changes

* minor changes

* removed from the upgraded service

* newline change

* Fix Part of #8472: Migrate Topic Rights Backend API Service (#11177)

* Migrate Topic Rights Backend API to angular

* Remove unecessary comments

* Fix minor issues

* Add rootScope

* Add .()

* issue fixed

* changed to 1s

* delay for mobile only

* only mobile

* added frontend tests

* backend tests fixed

* Update assets/i18n/en.json

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Update assets/i18n/en.json

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* fixed issues

* removed frontend test

* minor fixes

* fixed issues

* minor fixes

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>
Co-authored-by: Sandeep Dubey <dubeysandeep.in@gmail.com>
Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>
Co-authored-by: Kevin Thomas <kevintab@tutanota.com>
Co-authored-by: Prayush Dawda <35144226+iamprayush@users.noreply.github.com>
Co-authored-by: Ramin Izadpanah <ramizd@gmail.com>
Co-authored-by: Matt Higgins <mrhiggi@g.clemson.edu>
Co-authored-by: Matt Higgins <matt@matthiggins.dev>
Co-authored-by: arpit1912 <46648172+arpit1912@users.noreply.github.com>
Co-authored-by: Siddharth Batra <sbsiddharth@gmail.com>
Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

krishita30j committed:
Fixes issue #11830: Console error in the collection player screen (#11959)

* Modified files

* Modified files

* Fixed linting errors

* Modified files

* Fixing linting errors

* Modified files

* Fixing linting errors

vojtechjelinek committed:
Address comments from #11946 (#11981)

srijanreddy98 committed:
Fixes part of #9749: Migrates all the Algebraic Interactions to angular components. (#11958)

* Migrate Conitnue interaction

* Add codeowners

* Rename files in check frontend coverage

* Fix e2e?

* Add Continue to component list

* fix count of files

* Migrate algebraic interactions

* Remove extra new line

* Cover the file completely

* Review

* Review

17sushmita committed:
Added separate file interaction-defs for InteractionId type definition

17sushmita commented:
@iamprayush @DubeySandeep PTAL

U8NWXD committed:
Upgrade Redis CLI to 6.0.10 (#11994)

DubeySandeep commented on core/templates/services/interaction-specs.service.spec.ts:
```diff
@@ -29,6 +29,14 @@ describe('InteractionSpecsService', () => {
+      // of type '"ImageClickInput" | "NumberWithUnits" |
+      // "NumericExpressionInput" | "NumericInput" | "DragAndDropSortInput" |
+      // "ItemSelectionInput" | "Continue" | "GraphInput" | "EndExploration" |
+      // ... 11 more ... | "RatioExpressionInput"'." We need to suppress this
```
Don't use `11` as the number of interactions can be variable, (`...` is enough!)

DubeySandeep commented on core/templates/services/interaction-specs.service.ts:
```diff
@@ -20,10 +20,11 @@ import { downgradeInjectable } from '@angular/upgrade/static';
+  isInteractionTrainable(interactionId: InteractionId): boolean {
     const interactionSpecs =
       InteractionSpecsConstants.INTERACTION_SPECS[interactionId];
     if (!interactionSpecs) {
```
Do we still need this check?

17sushmita replied:
I don't think so, because we have now defined the type for interactionId. As I am a new contributor here, I was not sure to remove this. I'm making a commit to remove the check.

DubeySandeep commented on extensions/interactions/interaction-defs.ts:
```diff
@@ -0,0 +1,21 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
2021*

DubeySandeep reviewed (changes requested):
@17sushmita, This looks great, I've left a few minor comments PTAL!

Rijuta-s committed:
adding display and margin to bullet points (#11976)

mridul-netizen committed:
Fixed some minor changes for the Update-Audio-Translation window (#11956)

* added minor changes

* added some minor changes in update audio translation

* Update mark-all-audio-and-translations-as-needing-update-modal.directive.html

* Update mark-audio-as-needing-update-modal.directive.html

gp201 committed:
Admin navbar test (#11962)

* added frontend tests

* added frontend tests

* added tests

* Update

Co-authored-by: Kevin Thomas <kevintab@tutanota.com>

* update

* update

* update

* update

* Updated

Co-authored-by: Kevin Thomas <kevintab@tutanota.com>

U8NWXD committed:
Fix #11947 and #11993: Fix URLs Reported to E2E Logging Server and Report Branch Name (#11995)

* Fix URL construction for reporting e2e status

* Add branch to e2e logging metadata

* Fix flake_checker_test

* Address review comments

17sushmita committed:
Removed unnecessary checks and resolved minor changes requested

17sushmita commented:
Hi @DubeySandeep, on writing the correct interactionId in the frontend test, the frontend tests are failing.
```
Chrome Headless 88.0.4324.150 (Linux x86_64) Content translation manager service with custom INTERACTION_SPECS cases should replace translatable customization args FAILED
        Expected $.dummyCustArg.value[0].content._unicode = 'first' to equal 'fr 1'.
        Expected $.dummyCustArg.value[1].content._unicode = 'first' to equal 'fr 2'.
        Error: Expected $.dummyCustArg.value[0].content._unicode = 'first' to equal 'fr 1'.
        Expected $.dummyCustArg.value[1].content._unicode = 'first' to equal 'fr 2'.
            at <Jasmine>
            at UserContext.<anonymous> (core/templates/combined-tests.spec.js:80787:51)
            at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:510231:30)
            at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:509716:43)
```
Can you please guide me further?

krishnarao22 committed:
Adding waitFor.pageToFullyLoad() (#12004)

sachinchauhan2889 committed:
Fix part of #4057: added tests in image-upload-helper.service.spec.ts (#11991)

SAEb-ai committed:
Fix #11892:Added a specific height to the cards. (#11911)

* Added a specifc height to all the cards

* Added a specifc height to all the cards

* Added a specifc height to all the cards

* Addressed review comments

* Addresssed review comments

* Fixed Height Issues

Co-authored-by: Md Shahbaz Alam <laser39@LAPTOP-NU0NCS8R.localdomain>

FareesHussain committed:
Fixes part of #10798: Fixing E2E tests for accessibility.js (#11999)

* using action.js functions for accessibility.js

* reverting browser.actions().sendKeys()

* removing elementToBeClickable

mridul-netizen committed:
Fix part of #9749 Migrating  subtopic-viewer-navbar-pre-logo-action.component.ts to Angular (#11983)

* migrated subtopic-viewer

* updated

* minor changes

* lint-checks

safwan20 committed:
added the test file for stateTutorialFirstTimeService (#12007)

Co-authored-by: safwan <safwan.mansuri@peak.ai>

Commit:
Contributor dashboard bug fixes (#12008)

* fix contributor dashboard bugs

* fix lint issues

* address comments

jimbyo committed:
update waitFor to waitForTimeout (#11970)

ashutoshc8101 committed:
Fixing End-to-End Tests with action.js and waitFor.js for protractor_utils/users.js fixes part of #10798 (#11930)

* Fixing End-to-End Tests with action.js and waitFor.js for protractor_utils/users.js partly fixes (#10798)

Linting issue fixed

linting fix

* Requested changes are implemented

* action use is reverted and appropiate comments added

* Request changes

* E2E Test fixed

EricZLou committed:
add to upgradedServices (#12009)

Rijuta-s committed:
Skill editor console (#11989)

* fixing #11929

* fix linting issue

* minor changes

* fixing #11989

* fixing lint issues

* fixing lint issues

* fixing #11989

* fixing lint issues

* fixing lint issues

* fixing lint issues

* minor fix

* renamingvariables anfd functions

srijanreddy98 committed:
Add @darksun27 to co-own all the angular-migration related files (#12003)

shubhamkorde committed:
Fix of part of #10798 end-to-end test general.js file (#11965)

* fixed e2e tests using action.js and waitfor.js

* fixed linting issues

* made the suggested changes

* added waitfor to and from elements in drag and drop script

* changed presenceOf to visibility of

* changed container name

* added await statement

* changes the code according to style guide

iamprayush reviewed (approved):
LGTM for code-owner files!

srijanreddy98 committed:
Fix part of #9749: Migrate all fraction interactions (#11990)

Commit:
Update authors and changelog for v3.0.8 (#12025)

* Update CHANGELOG

* Update AUTHORS

* Update CONTRIBUTORS

* Update core/templates/pages/about-page/about-page.constants.ts

* Update package.json

* Remove blank entry; Undo unnecessary changes in package.json

* Fix names

brianrodri committed:
initial commit (#11973)

shubhamkorde committed:
Delete learner-answer-details-data service as it was dead code. (#11954)

* renamed changes

* changed imports

* removed focus from tests

* fixed linting issues

* refactored the code

* removed the httpclient imports from the data service

* Made the suggested changes

* deleted service as it was dead-code

vojtechjelinek committed:
Add emulator mode (#11906)

U8NWXD committed:
Fix #12036: Add Link to Wiki for Shards-Filesystem Mismatch (#12037)

subodhk01 committed:
Fixes part of #10306 : Added async keyword to asynchronous functions that return a promise (#12022)

* Async function updates for exploration-features-backend-api.service.ts

* Passed eslint checks

krishnarao22 committed:
Resolving the Error in Adding Screenshots (Github Actions) (#12014)

* Adding screenshots to one suite

* Switching position of uploading artifacts step

* Using env variable in path, adding debug lines

* fixing path

* Adding upload artifact step to all suites, removing debug lines

* Uploading artifacts even when tests fail

* Inducing failure, adding startsWith() to additionalEditor test

* Removing induced failure, adding startsWith() ot all suites

* Removing ${{ }}

ReshuKumari committed:
Placing the lesson image in center of lesson card and removing text with 0 story (#11951)

* placing the images in center in lesson cards

* lint error

* checking coming soon card

* minor change

* addressing review

* review changes

* correcting text shading for mobile view

* addressing review

* lint error

sachinchauhan2889 committed:
add async to topic-viewer-backend-api.service.ts (#12032)

Archi34 committed:
Add 'Classroom' option in the sidebar menu (mobile specific) with 'Basic Mathematics' option in submenu. (#11985)

* Add classroom option in sidebar

* Fix backend test

* Fix review changes

* Fix review changes

* Fix review changes

* Fix review changes

sachinchauhan2889 committed:
validators for PencilCodeEditor (#12033)

fix-lint

need to revert later

final-fix

fix

Nik-09 committed:
Fix Part of #10474 : Makes typescript check strict (#12058)

sajalasati commented:
> Hi @DubeySandeep, on writing the correct interactionId in the frontend test, the frontend tests are failing.
> 
> ```
> Chrome Headless 88.0.4324.150 (Linux x86_64) Content translation manager service with custom INTERACTION_SPECS cases should replace translatable customization args FAILED
>         Expected $.dummyCustArg.value[0].content._unicode = 'first' to equal 'fr 1'.
>         Expected $.dummyCustArg.value[1].content._unicode = 'first' to equal 'fr 2'.
>         Error: Expected $.dummyCustArg.value[0].content._unicode = 'first' to equal 'fr 1'.
>         Expected $.dummyCustArg.value[1].content._unicode = 'first' to equal 'fr 2'.
>             at <Jasmine>
>             at UserContext.<anonymous> (core/templates/combined-tests.spec.js:80787:51)
>             at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:510231:30)
>             at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:509716:43)
> ```
> 
> Can you please guide me further?

@17sushmita Please don't forget to assign the correct person while waiting for a review or a reply to avoid unwanted delays. Assigning @DubeySandeep as per your comment.

ReshuKumari committed:
Classroom page tasks adding topic in lesson player (#11923)

* task 2

* learner nav changes

* task 2

* failing e2e tests

* solving e2e tests

* fixing failing e2e

* fe error fix

* addressing review

seanlip committed:
Update fractions placeholder text. (#12055)

U8NWXD committed:
Fix Skipping E2E Tests for Changelog Update PRs (#12035)

On CircleCI, we weren't skipping E2E tests for changelog update PRs.
This commit fixes that, and in the process fixes #12026.

shubha-rajan committed:
Fix part of #11475: Migrate Base Model Validator to Apache Beam (#11955)

* Add clone method to BaseModel

* added beam PTransform for base model validator

* update docstrings

* update codeowners

* run autopep8

* run docformatter

* linting fixes

* linting fixes

* address review comments

* more linting fixes

* address review comments and move clone function

* Use error classes instead of tuples for output

* linting fixes

* address review comments

* linting

* update third party size check

* address review comments

* fix class name in tests

* move clone function into utils.py

* fix assert function

* address review comments

* fix linting

* Update jobs/base_model_validator.py

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Update jobs/base_model_validator.py

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* address review comments

* review comments

* fix base_model_validator_errors

* linting fixes

* Update jobs/base_model_validator.py

Co-authored-by: Brian Rodriguez <brian@brianrodri.com>

* Update jobs/base_model_validator.py

Co-authored-by: Brian Rodriguez <brian@brianrodri.com>

* Update base_model_validator.py

* Update base_model_validator_errors.py

* update base_model_validator_errors.py

* Update jobs/base_model_validator.py

Co-authored-by: Brian Rodriguez <brian@brianrodri.com>

* address review comments

* add jobs tests to shards

* Update backend_test_shards.json

* Update jobs/base_model_validator.py

Co-authored-by: Brian Rodriguez <brian@brianrodri.com>

* fix error in base_model_validator

* remove __init__ from ValidateModelIdWithRegex

* rename PERIOD_TO_HARD_DELETE_MODEL_IN_DAYS

* add regex_string to docstring

* fix linting errors

* add tests for ModelValidationError

* Update base_model_validator.py

* Update base_model_validator.py

* update message property in ModelValidatorError to raise NotImplementedError

Co-authored-by: Brian Rodriguez <brian@brianrodri.com>
Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

aishwary023 committed:
Added e2e test for student user journey (#12010)

* Added e2e tests

* Lint fix

* Addressed comments

* Fixed linting issues

* Addressed comments

* Lint checks

U8NWXD committed:
Fix #12029: Escape Protractor Screenshot File Names (#12034)

* Escape Protractor Screenshot File Names

GitHub Actions does not allow uploaded files to contain any of
`",:,<,>,|,*,?`. This commit configures the protractor screenshot
reporter to escape these characters. This fixes #12029.

* Intentionally introduce failure for testing

* Revert "Intentionally introduce failure for testing"

This reverts commit 8ea8f0c07eb7da2376d22c94b55fb0418a4f15e3.

ashutoshc8101 committed:
Fix part of #9749: Summary list header directive migrated to angular (#12019)

* Summary List Header migrated

* Linting fixed

* Update core/templates/components/state-directives/answer-group-editor/summary-list-header.component.ts

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Update core/templates/components/state-directives/answer-group-editor/summary-list-header.component.ts

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Suggested changes

* Fixed indentation

* Refactered

* changes

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

srijanreddy98 committed:
Fix part of #8472: Migrate image-upload-helper.service (#11306)

* Migrate Service File

* Migrate Spec File

* Add Service Imports

* add component for thumbnail display

* Add component.ts

* Add component

* add in shared components

* Fix Non display of images when uploading

* Fix Linting Issues

* Add file to not tested

* Add file to not tested

* Migrate image uploader helper service

* Migrate image uploader helper service

* Migrate image uploader helper service

* Review

* Add new service

* Add svg sanitizer to the frontend coverage list

* Add svg sanitizer to the frontend coverage list

* Review

* Review

* Review

* Review

* Review

* Review

* Changes

* Changes

* Review

* Add codeowners

* Review

* Changes

* review

* Fix unit tests

* Fix unit tests

* Merge with upstream/develop

* Update tests

* Remove svg-sanitizer from denylist

Co-authored-by: darksun27 <sbsiddharth@gmail.com>

seanlip committed:
Create FUNDING.yml

vojtechjelinek committed:
Fix #11431 and fix part of #8510: Clean-up unneeded code (#12051)

DubeySandeep commented:
@17sushmita Sorry for the late reply, can you please let me know the correct InteractionId you have used in the test?

ashutoshc8101 committed:
User email preferences service migration from angularjs to angular fixes part of (#8472) (#11949)

* Migration complete

* Two left

* Complete

* Changed to backend api service

* Antoher one

* another one

* sd

* sfd

* sf

* Final

* Final 2

* sdf

* Wrong implementation reverted

* Refactored

* File overview added

* undo changes to settings-tab

* undo settings-tab 2

* undo exploration-editor-page

* undo setting tab 3

* Requested changes are resolved

* Tests are fixed

* Fixed

* callback

* fixed setting-tab.component.spec.ts

* Suggested changes

* sorted

* Removed the empty line

* requested changes by srijanreddy

* Requested changs

* changes

* sdf

* refactor

* fremove f

* 100% coverge for settings-tab.component.ts

* f removed

* requested changes

* added empty lines

* changes

brianrodri committed:
Fix part of #11462: Transition backend to manage Firebase authentication using HTTP cookies (#12054)

* initial commit

* cleanup

* fix emulator

* address comments

* fix unrelated lint errors

* Merge remote-tracking branch 'upstream/develop' into fix-backend-firebase-usages

* Revert "Merge remote-tracking branch 'upstream/develop' into fix-backend-firebase-usages"

This reverts commit e7b71ae1424bfa31ef3a1a9af6118316dd987a4c.

* address comments

* address comments

* fix tests

* address comments

* remove unused function

* address comments

* address comments

* address comments

* address comments

* address missing comment

krishita30j committed:
Fix #10919, #8800 & #4932: Making "Play later" section tiles consistent with other sections (#12048)

17sushmita commented:
> @17sushmita Sorry for the late reply, can you please let me know the correct InteractionId you have used in the test?

@DubeySandeep I used `ImageClickInput` in place of `DummyInteraction` for the interactionId.

17sushmita commented:
@DubeySandeep PTAL

DubeySandeep commented on core/templates/pages/exploration-player-page/services/content-translation-manager.service.spec.ts:
```diff
@@ -322,6 +324,14 @@ describe('Content translation manager service', () => {
+      // @ts-expect-error
       interaction.id = 'DummyInteraction';
       interaction.customizationArgs = {
         dummyCustArg: {value: [{
```
@shavavo, Can you please confirm  that the `unicode_str` on line 358 and 365 should be "first"? (line number are from the committed file)

DubeySandeep replied:
@17sushmita It looks like this service doesn't work for some interaction (Which doesn't have any translatable contents) can you please add a proper custArg and interaction id so that it uses the translated text?

[ref](https://github.com/oppia/oppia/blob/4df7eadfd73124d58e5d2eb5223160061ab32fe4/core/templates/pages/exploration-player-page/services/content-translation-manager.service.ts#L167)

atpug22 committed:
Fix part of #4057: Added a unit test for Graph Utils Service (#12072)

* commit

* covered full graph-utils.service.ts

* fixed one lint check

* minor change in name

Radesh-kumar committed:
Fix part of #8016: Added a separate backend api service for admin-jobs-tab.directive.ts (#11446)

* start-new-job

* frontend tests

* lint checks

* updated

* updated

* updated

* updated

* updated

* updated

* updated

* updated

* update

* Update admin-jobs-tab.directive.ts

* final update

* final update

* final update

* final update

* final update

* fixed e2e tests

* updated

* e2e updated

* finished

* final final

sagangwee committed:
Fix #12061: Add email message check to resolve server error when submitting question suggestions (#12080)

* Add email checks

* Remove unreachable defensive check

seanlip committed:
Clear old drafts and fix format of content suggestions. (#12068)

* Temporary commit to record investigation jobs.

* Revert "Temporary commit to record investigation jobs."

This reverts commit a39d79dc97438c65d4b25a536833de01271ff281.

* Quality-of-life improvements.

* Add new jobs.

* Add tests and address review comments.

Commit:
Fix #11441: Fixes a frontend flake (#12064)

* fix flake

* remove unnecessary import

brianrodri committed:
only start firebase emulator in EMULATOR_MODE (#12093)

YBCS committed:
Fix part of #10306: Add async to `story-viewer-backend-api.service.ts` (#12081)

Commit:
Preload functionality for images in state writtenTranslations (#12063)

* Preload functionality for images in state writtenTranslations

* fix typescript error

* Updates to UpgradedServices.ts

* fix for customisationArgs in certain interactions

* fix frontend failure

amulya00 committed:
Fixes Part of #10700: Refactor Suggestions object factory (#11942)

* Fix part of #10306: Added async keyword to functions that return a promise

* added async keyword to function getExplorationStatsAsync

* Update exploration-improvements-task-registry.service.ts

* Made necessary changes

* removed unnecessary newline

* Fix Part of #10700: Refactor suggestion object factory

* Fix Part of #10700: Refactor suggestion object factory

* Fix Part of #10700: Refactor suggestion object factory

* Fix Part of #10700: Refactor suggestions object factory

* Fix Part of #10700: Refactor suggestion object factory

* Fix Part of #10700: Refactor suggestion object factory

* Fixed indentation in suggestion.model.ts

* Fixed indentation in suggestion.model.ts

* Fixed indentation in suggestion.model.ts

* Fix Part of #10700: Refactor suggestion object factory

* Fixed suggestion.model.ts

* Fixed suggestion.model.ts

* Fixed typerscript errors

* Fixed typerscript errors

* Fixed typerscript errors

* Fixed typerscript errors

* Fixed typerscript errors

* typerscript errors

* typerscript errors

* typerscript errors

* typerscript errors

sachinchauhan2889 committed:
Adds correct filename for the script checking frontened coverage (#12012)

AdityaDubey0 committed:
Fix css linter script issue (#12105)

seanlip committed:
Remove suggestion submission functionality from the learner view. (#12096)

* Remove suggestion submission functionality from the learner view.

* Disallow content edit suggestions from being submitted to the backend.

* Fix backend tests.

brianrodri committed:
Fix part of #11462: Add one-off job to populate Firebase servers (#12086)

* initial commit

* cleanup

* fix emulator

* address comments

* fix unrelated lint errors

* Merge remote-tracking branch 'upstream/develop' into fix-backend-firebase-usages

* Revert "Merge remote-tracking branch 'upstream/develop' into fix-backend-firebase-usages"

This reverts commit e7b71ae1424bfa31ef3a1a9af6118316dd987a4c.

* address comments

* address comments

* fix tests

* address comments

* remove unused function

* address comments

* address comments

* address comments

* initial commit

* address comments

* address missing comment

* add email audit

* fix lint

* address comments

* fix typos

* address comments

* address comments

* add to test shards

* address comments

* address comments

* address SYSTEM_COMMITTER_ID

* add info for SYSTEM_COMMITTER_ID

AdityaDubey0 committed:
Fix part of #5097: Add length-zero-no-unit rule for CSS lint check (#12095)

* Add new rules for CSS lint checks

* Added correct indentetion check

* Remove indentation check and fixes length-zero-no-unit rules for CSS lint checks

ashutoshc8101 committed:
Fix part of #8472: collection-editor-state.service.ts along with spec migrated (#12027)

* worked

* lint fixes

* adds

* Update core/templates/pages/collection-editor-page/services/collection-editor-state.service.ts

Co-authored-by: Siddharth Batra <sbsiddharth@gmail.com>

* requested changes by darksun

* COllectionChange removed

* requested changes

Co-authored-by: Siddharth Batra <sbsiddharth@gmail.com>

vojtechjelinek committed:
Introduce wipeout related utils to admin page (#12101)

ARKEOLOGIST committed:
Fix part of #11919 : Added styles to prevent errors from looking clickable (#12039)

* Fixing errors being clickable on editor pages

* Fixing issues in skill editor

brianrodri committed:
Improve Elasticsearch process management (#12111)

Radesh-kumar committed:
Fix part of #4057: Added a unit test for Hint-and-solution-modal Service (#12094)

* updated

* updated

* updated

* updated

* fixed

* Update core/templates/pages/exploration-player-page/services/hint-and-solution-modal.service.spec.ts

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

brianrodri committed:
Run Elasticsearch server in quiet mode (#12123)

Nik-09 committed:
removes todo 10110 comments (#12112)

Co-authored-by: nikhil <Nik-09>

jimbyo committed:
Fix #10717 Surfacing translation meta deta (#12047)

* Resurface metadata

* Fix injector error

* Fix lint

* fix unit tests

* fix lints

* test subheading

hardikkat24 committed:
Fix part of #10049: Remove deprecated fields from ExplorationModel (#11998)

* fix part of #10049: remove skill_tags, default_skin and skin_customizations fields

* Add _reconstitute test and addressed PR comments

* changes

* added job to jobs_registry

* split test into 3 tests

* lint errors fixed

* changes

* MapReduce Error resolved

* changes

* lint issues fixed

* _reconstitute replaced by get_version

* update_timestamp changed

* job improved and comments removed

* _reconstitute improved

* changes

* changes-lint

* changes-lint

* bugs fixed

* temp file deleted

* lint-issues

* Changed date

* wrong comments removed

* using dict instead of model

Co-authored-by: Michael Wu <michaelywu54@gmail.com>

17sushmita commented:
@DubeySandeep I tried to use 'TextInput' interactionId which uses translated text but that also didn't work. But if you see the error it says that 'Content translation manager service with `custom INTERACTION_SPECS cases` should replace translatable customization args FAILED' , which means this test expects custom args with custom INTERACTION_SPECS cases. It should work for custom interactionId only I think and not correct InteractionId. What do you think?
![image](https://user-images.githubusercontent.com/42088159/110251917-e8173380-7fa8-11eb-8137-faacb638adf4.png)

@DubeySandeep PTAL

DubeySandeep commented:
@17sushmita, Thanks for checking, I agree, let's keep the test as-is i.e, with invalid interaction-id!

U8NWXD commented:
@17sushmita there seem to be a lot of unrelated commits that got added recently, along with a bunch of new codeowners. Was this intentional?

DubeySandeep commented:
@17sushmita, Looks like there's some issue with your last commit, PTAL!

17sushmita commented:
> @17sushmita, Looks like there's some issue with your last commit, PTAL!

Hi @DubeySandeep , I did git pull upstream and then push the changes which pulled these recent changes. Now, to go back to the commit I need to force push but the guideline says forced push will close the PR.What should I do?

DubeySandeep commented:
> Hi @DubeySandeep , I did git pull upstream and then push the changes which pulled these recent changes. Now, to go back to the commit I need to force push but the guideline says forced push will close the PR.What should I do?

Can you please create another PR following the wiki guidelines? :) [I'm sorry, I don't know how this happened but do let me know if I can help anyway?]

DubeySandeep commented:
@17sushmita I'm closing this PR as it's not in a state to review. Can you please create a new PR following the wiki guidelines with all the expected changes? (sorry!)
